### PR TITLE
N3 operation retractions

### PR DESCRIPTION
**Pull request by VeljkoMaksimovic**
Extended N3TemplateOperation (both definition and implementation) to also support retraction of triplets.

Question
How to delete multiple entries with one n3 triplet? For example, delete all statements that start with <http://Joe> <http://Has>

### CODE REVIEW

**Comment by chenejac** [comment]
> For now, both n3TextAdditions and n3TextRetraction property cardinalities are set to min:0, max:1. That allows to have n3Template that has no addition statements, nor retraction statements, which should not be valid. Not sure how to set "collective cardinality" on those two properties to be min:1

Probably this is a task for me to be resolved by definition of a SHACL rule.

**Comment by litvinovg** [comment]
I think we need integration tests to check that action with n3 template is converted into java action with configurationBeanLoader and instantiated action with n3 template can process operation data.

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/N3Template.java [hunk 1]
```diff
@@ -19,7 +19,8 @@ public class N3Template extends Operation implements Template {
 
 
 	private Parameters requiredParams = new Parameters();
-	private String n3Text;
+	private String n3TextAdditions;
```
Comment (litvinovg) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/N3Template.java:22]: Maybe initialize empty string to get rid of null checks?

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/N3Template.java [hunk 2]
```diff
@@ -61,31 +69,35 @@ public OperationResult run(OperationData input) {
 			return new OperationResult(500);
 		}
 
-		String substitutedN3Template;
+		String substitutedN3AdditionsTemplate = null, substitutedN3RetractionsTemplate = null;
 		try {
-			substitutedN3Template=insertParameters(input);
+			if (this.n3TextAdditions != null) substitutedN3AdditionsTemplate=insertParameters(input, this.n3TextAdditions);
```
Comment (litvinovg) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/N3Template.java:74]: I would suggest to always use curly braces

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/N3Template.java [hunk 3]
```diff
@@ -61,31 +69,35 @@ public OperationResult run(OperationData input) {
...
+		List<Model> additionModels = new ArrayList<>(), retractionModels = new ArrayList<>();
 		try {
-			additionModels = ProcessRdfForm.parseN3ToRDF(Arrays.asList(substitutedN3Template), ProcessRdfForm.N3ParseType.REQUIRED);
+			if (substitutedN3AdditionsTemplate != null)
+				additionModels = ProcessRdfForm.parseN3ToRDF(Arrays.asList(substitutedN3AdditionsTemplate), ProcessRdfForm.N3ParseType.REQUIRED);
+			if (substitutedN3RetractionsTemplate != null)
+				retractionModels = ProcessRdfForm.parseN3ToRDF(Arrays.asList(substitutedN3RetractionsTemplate), ProcessRdfForm.N3ParseType.REQUIRED);
 		} catch (Exception e) {
 			log.error("Error while trying to parse N3Template string and create a Jena rdf Model");
 			log.error(e);
```
Comment (litvinovg) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/N3Template.java:89]: I would suggest log.error(e,e);

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/N3Template.java [hunk 4]
```diff
@@ -61,31 +69,35 @@ public OperationResult run(OperationData input) {
...
 			return new OperationResult(500);
 		}
 
-		List<Model> additionModels;
+		List<Model> additionModels = new ArrayList<>(), retractionModels = new ArrayList<>();
 		try {
-			additionModels = ProcessRdfForm.parseN3ToRDF(Arrays.asList(substitutedN3Template), ProcessRdfForm.N3ParseType.REQUIRED);
+			if (substitutedN3AdditionsTemplate != null)
+				additionModels = ProcessRdfForm.parseN3ToRDF(Arrays.asList(substitutedN3AdditionsTemplate), ProcessRdfForm.N3ParseType.REQUIRED);
```
Comment (litvinovg) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/N3Template.java:84]: curly braces

**Review by litvinovg — CHANGES_REQUESTED** [review]
It makes sense to rename n3Template to N3Template in ontology as you fixing this.

**Comment by brianjlowe** [comment]
Agree with all of Georgy's requested changes.  In addition, watch line length.  I usually try to not to exceed 80 characters (which may be more conservative than necessary), but as it currently some of the lines needing curly braces overflow the div and are hidden when using Github's Show File view.  This suggests that they are simply too long.

When Georgy's satisfied with the changes, I can add my second approval.  While I don't fully understand the context or use cases, in isolation the approach of adding a retractions template and processing it via the same mechanism as additions seems logical enough.

**Comment by VeljkoMaksimovic** [comment]
> Agree with all of Georgy's requested changes. In addition, watch line length. I usually try to not to exceed 80 characters (which may be more conservative than necessary), but currently some of the lines needing curly braces overflow the div and are hidden when using Github's Show File view. This suggests that they are simply too long.
> 
> When Georgy's satisfied with the changes, I can add my second approval. While I don't fully understand the context or use cases, in isolation the approach of adding a retractions template and processing it via the same mechanism as additions seems logical enough.

Ok, thanks for the suggestion, will do later in the day

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/N3Template.java [hunk 5]
```diff
@@ -61,31 +68,33 @@ public OperationResult run(OperationData input) {
 			return new OperationResult(500);
 		}
 
-		String substitutedN3Template;
+		String substitutedN3AdditionsTemplate = "";
+		String substitutedN3RetractionsTemplate = "";
 		try {
-			substitutedN3Template=insertParameters(input);
+			n3TextAdditions = substitutedN3AdditionsTemplate=insertParameters(input, n3TextAdditions);
```
Comment (litvinovg) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/N3Template.java:74]: formatting + n3TextAdditions shouldn't be changed during execution.

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/N3Template.java [hunk 6]
```diff
@@ -61,31 +68,33 @@ public OperationResult run(OperationData input) {
...
+			n3TextRetractions = substitutedN3RetractionsTemplate=insertParameters(input, n3TextRetractions);
```
Comment (litvinovg) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/N3Template.java:75]: formatting + n3TextRetractions shouldn't be changed during execution.

**Review by litvinovg — CHANGES_REQUESTED** [review]

**Comment by litvinovg** [comment]
Please add test to verify that after one run of n3template with input data second run with different input data produces different results.

**Comment by VeljkoMaksimovic** [comment]
> Please add test to verify that after one run of n3template with input data second run with different input data produces different results.

Ok, I added one extra test that does that.

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/N3Template.java [hunk 7]
```diff
@@ -61,31 +68,38 @@ public OperationResult run(OperationData input) {
...
-			substitutedN3Template=insertParameters(input);
+			substitutedN3AdditionsTemplate = insertParameters(input, n3TextAdditions);
+			substitutedN3RetractionsTemplate = insertParameters(input, n3TextRetractions);
 		}catch (InputMismatchException e){
 			log.error(e);
 			return new OperationResult(500);
 		}
 
-		List<Model> additionModels;
+		List<Model> additionModels, retractionModels;
```
Comment (litvinovg) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/N3Template.java:81]: Please declare variables on separate lines

**Review by litvinovg — APPROVED** [review]
Looks good to me.

**Review by brianjlowe — APPROVED** [review]

**Pull request merged**